core/src/main/java/kafka/server/SharePartition.java

```diff
@@ -280,7 +294,7 @@ public long nextFetchOffset() {
         try {
             // When none of the records in the cachedState are in the AVAILABLE state, findNextFetchOffset will be false
             if (!findNextFetchOffset.get()) {
-                if (cachedState.isEmpty()) {
+                if (cachedState.isEmpty() || startOffset > cachedState.lastEntry().getValue().lastOffset) {
```

Comment:
Can there be a possibility that `endOffset` is also prior to `startOffset`?

Reply:
yes, we handle that case within `updateOffsetsOnLsoMovement()` and to correctly calculate the nextFetchOffset, we need this change.